ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -401,167 +455,369 @@ void PowerSystemNode::injectCustomFault(bool& fault_activated,
+    // simply update wattage based on the current value of the HPD slider.
+    if (fault_name == FAULT_NAME_HPD_ACTIVATE)
+    {
+      ros::param::getCached("/faults/" + FAULT_NAME_HPD, hpd_wattage);
+      double split_wattage = hpd_wattage / NUM_MODELS;
+
+      for (int i = 0; i < NUM_MODELS; i++)
+      {
+        m_power_models[i].model.setHighPowerDraw(split_wattage);
+      }
+    }
+
+    // NOTE: If other faults are added to the RQT window in the future,
+    // this is where their related flags/logic would be checked/used
+    // in a similar manner to the above code block with high power draw.
+  }
+}
+
+/*
+ * Basic function that calls the other fault-handling functions.
+ */
+void PowerSystemNode::injectFaults()
 {
   injectFault(FAULT_NAME_HPD_ACTIVATE,
-              m_high_power_draw_activated,
-              power, voltage, temperature);
+              m_high_power_draw_activated);
   injectCustomFault(m_custom_power_fault_activated,
                     m_custom_power_fault_sequence,
-                    m_custom_power_fault_sequence_index,
-                    power, voltage, temperature);
+                    m_custom_power_fault_sequence_index);
 }
 
-PrognoserMap
-PowerSystemNode::composePrognoserData(double power,
-				      double voltage,
-				      double temperature)
+/*
+ * Callback function that publishes mechanical power values and sends
+ * mechanical power, which is converted into power draw, to each model in the
+ * node. The mechanical power is distributed evenly amongst all models.
+ */
+void PowerSystemNode::jointStatesCb(const sensor_msgs::JointStateConstPtr& msg)
 {
-  return PrognoserMap {
-    { MessageId::Watts, Datum<double>{ power } },
-    { MessageId::Volts, Datum<double>{ voltage } },
-    { MessageId::Centigrade, Datum<double>{ temperature } }
-  };
+  m_power_watts = 0.0;  // This includes the arm + antenna
+  for (int i = 0; i < ow_lander::NUM_JOINTS; ++i)
+    m_power_watts += fabs(msg->velocity[i] * msg->effort[i]);
+
+  m_power_values[m_power_values_index++ % m_power_values.size()] = m_power_watts;
+  m_mean_mechanical_power = accumulate(
+      begin(m_power_values), end(m_power_values), 0.0)
+      / m_power_values.size();
+
+  // Publish the total raw & average mechanical power.
+  std_msgs::Float64 mechanical_power_raw_msg, mechanical_power_avg_msg;
+  mechanical_power_raw_msg.data = m_power_watts;
+  mechanical_power_avg_msg.data = m_mean_mechanical_power;
+  m_mechanical_power_raw_pub.publish(mechanical_power_raw_msg);
+  m_mechanical_power_avg_pub.publish(mechanical_power_avg_msg);
+
+  // Send in mechanical power to each model, distributed evenly, to be converted
+  // into power draw.
+  for (int i = 0; i < NUM_MODELS; i++)
+  {
+    m_power_models[i].model.applyMechanicalPower(
+                                          (m_mean_mechanical_power / NUM_MODELS),
+                                          m_processing_power_batch);
+  }
 }
```

Comment:
This function pulls off a lot at 50Hz, and it does not need to. Assuming we are okay with mech power average/raw publishing at the same rate as the other battery topics instead of 50Hz (at the moment nothing subscribes to it), I suggest this function merely updates `m_power_values`.

There is no need for `applyMechanicalPower`, because the mechanical power is not applied at 50Hz, it's applied everytime `runOnce` is called. So I suggest, ahead of the loop where `runOnce` is called for each model handler, we call accumulate and pass the computed average power through `runOnce`. This make obsolete the `applyMechanicalPower` function and the semaphore boolean flags in `PrognoserInputHandler`.

Reply:
While I can get to work on this, iIt's worth noting that the two mechanical power topics are currently working as intended. They're supposed to publish every mechanical value calculated in the callback function (``raw``) as well as the ``average`` of the last several values according to the moving average window size.

Since the average is only used at the same rate as the main system loop, I can see ``mechanical_power/average`` being removed here to sync with the main loop instead, but if I update the ``raw`` publication to only publish at the rate of the main loop then it will only publish 1 out of every 50 raw values. This was the problem with the previous jointStatesCb implementation and is why I fixed it in OW-1187.

Reply:
I like the idea of still publishing raw at 50Hz, it's calculated so we may as well, but I still have performance concerns with the accumulate call at this rate. 

To show how much of an impact it has, I  deactivated the following lines of code in PowerSystemNode.cpp with comments and recompiled. 
```
  // m_joint_states_sub = m_nh.subscribe(
  //                       "/joint_states",
  //                       1,
  //                       &PowerSystemNode::jointStatesCb,
  //                       this);
```

The following plots are screen grabs from [KSysGuard](https://apps.kde.org/ksysguard/), and the simulator was running for their entire width. 

### Baseline
With the above code running, which results in the joinStatesCb called at 50Hz in up to 4 different threads at a time, I saw the following usage graph.
![with_mech_power_calc](https://github.com/nasa/ow_simulator/assets/17039973/31cb3654-2c66-4574-b780-30eb90733ba4)

### jointStatesCb deactivated
With the above code deactivated, jointStatesCb is never called and there is no average mech calculation performed, I saw this usage.
![without_mech_power_calc](https://github.com/nasa/ow_simulator/assets/17039973/6fa871b6-cd00-465d-be87-c21ef3c631d9)

### Caveats
Using KSysGuard is by no means the best way to perform this kind of analysis, but in comparison I think the results speak for themselves. **Note**: I consider the total grey area under both graphs to be more indicative of a higher CPU usage than the peaks and CPU switching behavior seen in the first graph

Deactivating jointStatesCb is not what I am suggesting, this is just a shorthand to show the performance impact of that one function. In reality, my suggestion would still have a vector of length 100 accumulated at 1 Hz (instead of 50Hz), and we cannot forget about the vector mutation, power calculation, and ROS topic API calls that would still occur at 50Hz. All of that overhead is not represented in the second graph.

### Conclusion
I am suggesting that the majority difference in CPU usage between these graphs is due to the accumulation of a 100 length vector at a rate of 50 Hz.

Reply:
There is a way to maintain publishing of the average at 50Hz but with greatly improved performance. Everytime accumulate is called, the middle values of the vector are added together, even if the majority of them were already added together in the previous calls. 

One way to do a moving average more efficiently is to use a "moving sum" and a FIFO array (m_power_values). Each frame you toss the oldest value out of the array, subtracting it from the **moving sum**, and append the newest value to the array, adding it to the **moving sum**. 

We have an example of something similar in our codebase, only it is a moving max filter, so rather than a sum it evaluates whether the newest value is larger than the current max value. [See here.](https://github.com/nasa/ow_simulator/blob/OCEANWATER-1069_balovnev_improvements/ow_gazebo_plugins/src/BalovnevModelPlugin/MovingMaxFilter.cpp) Note that it exists in a branch that is not yet merged into noetic-devel.

Reply:
Neat; I'll look into this and get to work on implementing it. Thanks for the references!

I didn't realize the performance impact ``accumulate`` had here, and since up until now ``jointStatesCb`` was running at 1/50th the rate it was supposed to, it was never caught earlier.

Reply:
My _hypothesis_ is that the performance impact is largely from accumulate, but my suggestion will simply have to be implemented before we can know for sure. Let me know when you've completed your changes and I would be happy to redo the same analysis 😄

Reply:
I'll prioritize it for the next commit, but I probably won't finish it until tomorrow (assuming no unexpected hurdles). I'll notify you on Teams when I do.

Even if it ends up not being the culprit, it's still a good exercise in writing more efficient code, so I'm happy to make the change.

Reply:
Well you were right to think something was off about this performance comparison. I did some significantly more thorough performance tests, and it turns out the new moving sum change does not reward us the performance boost that I promised. 

What happened in my first analysis was the prognosers weren't running at all, so no wonder performance was significantly better. This happened simply by unbinding `jointStatesCb` to the `/joint_states` method, but in commit 8d661ed this would result in a process flag never becoming true, resulting in prognosing never occurring. 

I still suspect we're seeing some much much smaller performance boost by avoiding the accumulate call, so I still think this was a worthwhile change. I just wanted to correct my earlier comment with the plots. 

Using a tool like valgrind would have been a better fit for this analysis, and I did try to do that this second time around by following the [process ROS provides](http://wiki.ros.org/roslaunch/Tutorials/Profiling%20roslaunch%20nodes), but had no luck getting it to work. The power_system_node either crashes or suspends when I prefix the valgrind command in the roslaunch. Unfortunate because this would be a very powerful tool to profile the simulation with.